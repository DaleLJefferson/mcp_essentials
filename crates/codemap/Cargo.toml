[package]
name = "codemap"
version.workspace = true
edition.workspace = true

[dependencies]
tree-sitter = "0.25.3"
tree-sitter-rust = "0.23.2"
ignore = "0.4.23"
tokio = { version = "1", features = ["full"] }

[dev-dependencies]
insta = "1"
