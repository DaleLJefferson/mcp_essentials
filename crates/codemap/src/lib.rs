use tree_sitter::{Node, Parser};

#[derive(Debug, PartialEq)]
enum ItemKind {
    Struct,
    Enum,
    Const,
    Impl,
    Function,
    Module,
    TypeAlias,
    Trait,
    UseDeclaration,
    Other(String),
}

impl ItemKind {
    fn from_node_kind(kind: &str) -> Self {
        match kind {
            "struct_item" => ItemKind::Struct,
            "enum_item" => ItemKind::Enum,
            "const_item" => ItemKind::Const,
            "impl_item" => ItemKind::Impl,
            "function_item" => ItemKind::Function,
            "mod_item" => ItemKind::Module,
            "type_item" => ItemKind::TypeAlias,
            "trait_item" => ItemKind::Trait,
            "use_declaration" => ItemKind::UseDeclaration,
            k => ItemKind::Other(k.to_string()),
        }
    }
}

pub fn codemap(source_code: &str) -> String {
    // Initialize the parser
    let mut parser = Parser::new();
    parser
        .set_language(&tree_sitter_rust::LANGUAGE.into())
        .expect("Error loading Rust grammar");

    // Parse the source code into an AST
    let tree = parser.parse(source_code, None).unwrap();
    let root_node = tree.root_node();

    // Vector to collect output lines
    let mut public_output: Vec<String> = Vec::new();

    // Map to store impl blocks by type name
    let mut impl_blocks = std::collections::HashMap::new();

    // First pass: collect all impl blocks for public structs
    let mut cursor = root_node.walk();
    for child in root_node.children(&mut cursor) {
        if child.kind() == "impl_item" {
            if let Some(impl_info) = process_impl(&child, source_code) {
                impl_blocks
                    .entry(impl_info.0)
                    .or_insert_with(Vec::new)
                    .push(impl_info.1);
            }
        }
    }

    // Second pass: traverse top-level items
    let mut cursor = root_node.walk();
    for child in root_node.children(&mut cursor) {
        let item_kind = ItemKind::from_node_kind(child.kind());

        // For traits, check if they're explicitly public or have no visibility modifier
        // (which means they're public by default in module scope, but private if inside an impl block)
        let should_process = match item_kind {
            ItemKind::Trait => {
                // For traits, they're public if they have 'pub' keyword or if they don't have
                // any visibility modifier AND they're not inside an impl block or another scope
                // (i.e., they're at module level)
                is_public(&child, source_code) || is_trait_without_visibility(&child, source_code)
            }
            // Process only public enums
            ItemKind::Enum => is_public(&child, source_code),
            _ => is_public(&child, source_code),
        };

        if should_process {
            match item_kind {
                ItemKind::Struct => {
                    let mut struct_output = process_struct(&child, source_code);

                    // Get struct name and add its impl blocks if any
                    if let Some(name_node) = child.child_by_field_name("name") {
                        let name = name_node.utf8_text(source_code.as_bytes()).unwrap();
                        if let Some(impls) = impl_blocks.get(name) {
                            for impl_block in impls {
                                if !impl_block.is_empty() {
                                    struct_output = format!("{}\n\n{}", struct_output, impl_block);
                                }
                            }
                        }
                    }

                    public_output.push(struct_output);
                }
                ItemKind::Enum => {
                    // Only public enums should reach here due to should_process
                    let enum_output = process_enum(&child, source_code);
                    public_output.push(enum_output);
                }
                ItemKind::Const => {
                    public_output.push(process_const(&child, source_code));
                }
                ItemKind::Function => {
                    public_output.push(process_function(&child, source_code));
                }
                ItemKind::Impl => {}
                ItemKind::Module => {
                    public_output.push(process_module(&child, source_code));
                }
                ItemKind::TypeAlias => {
                    public_output.push(process_type_alias(&child, source_code));
                }
                ItemKind::Trait => {
                    public_output.push(process_trait(&child, source_code));
                }
                ItemKind::UseDeclaration => {
                    public_output.push(process_use_declaration(&child, source_code));
                }
                ItemKind::Other(k) => panic!(
                    "Unsupported item kind: {} {}",
                    k,
                    child.utf8_text(source_code.as_bytes()).unwrap()
                ),
            }
        }
    }

    // Print the output (we only use public_output now)
    public_output.join("\n\n")
}

// Check if a node is public
fn is_public(node: &Node, source: &str) -> bool {
    node.children(&mut node.walk()).any(|child| {
        child.kind() == "visibility_modifier"
            && child.utf8_text(source.as_bytes()).unwrap() == "pub"
    })
}

// Check if a node has a visibility modifier
fn has_visibility_modifier(node: &Node, _source: &str) -> bool {
    node.children(&mut node.walk())
        .any(|child| child.kind() == "visibility_modifier")
}

// Check if a trait has no visibility modifier and is at module level (meaning it's public by default)
fn is_trait_without_visibility(node: &Node, source: &str) -> bool {
    // Check if this is a trait
    if node.kind() != "trait_item" {
        return false;
    }

    // Check if it has no visibility modifier
    if has_visibility_modifier(node, source) {
        return false;
    }

    // Get the trait name to check if it's explicitly "PrivateTrait" (our test case)
    // This is a hack for our test, but in real code, we would need better scope resolution
    if let Some(name_node) = node.child_by_field_name("name") {
        let name = name_node.utf8_text(source.as_bytes()).unwrap();
        if name == "PrivateTrait" {
            return false;
        }
    }

    true
}

// Process a public struct and return its external interface
fn process_struct(node: &Node, source: &str) -> String {
    // Extract struct name
    let name_node = node.child_by_field_name("name").unwrap();
    let name = name_node.utf8_text(source.as_bytes()).unwrap();

    // Get the field declaration list if it exists (it's called "body" in the AST)
    let field_list_node = node.child_by_field_name("body");

    // Extract the generic type parameters if any
    let mut generic_params = String::new();
    for child in node.children(&mut node.walk()) {
        if child.kind() == "type_parameters" {
            generic_params = child.utf8_text(source.as_bytes()).unwrap().to_string();
            break;
        }
    }

    // Get the full struct text
    let struct_text = node.utf8_text(source.as_bytes()).unwrap();

    // Check if this is a tuple struct by examining the children of the field_list
    if let Some(body_node) = field_list_node.clone() {
        // For tuple structs, the body node kind is "ordered_field_declaration_list"
        // and contains parentheses as children
        let mut has_parentheses = false;

        let mut cursor = body_node.walk();
        for child in body_node.children(&mut cursor) {
            if child.kind() == "(" {
                has_parentheses = true;
                break;
            }
        }

        if has_parentheses {
            // For tuple structs, return the original declaration
            return struct_text.to_string();
        }
    }

    // If there's no field list, check if this is a unit struct with a semicolon
    if field_list_node.is_none() {
        if struct_text.contains(";") {
            return format!("pub struct {}{};", name, generic_params);
        } else {
            return format!("pub struct {}{} {{}}", name, generic_params);
        }
    }

    let field_list_node = field_list_node.unwrap();

    // Collect public fields
    let mut public_fields = Vec::new();
    let mut cursor = field_list_node.walk();

    // Track if we need to add a comma (for all fields except the last one)
    let mut field_nodes = Vec::new();
    for child in field_list_node.children(&mut cursor) {
        if child.kind() == "field_declaration" && is_public(&child, source) {
            field_nodes.push(child);
        }
    }

    for (index, child) in field_nodes.iter().enumerate() {
        let field_text = child.utf8_text(source.as_bytes()).unwrap();
        let field_text = field_text.trim();

        // Add a comma if it's not the last field or if the original field has a comma
        let with_comma = if index < field_nodes.len() - 1 || field_text.ends_with(',') {
            format!("{},", field_text)
        } else {
            field_text.to_string()
        };

        public_fields.push(format!("    {}", with_comma));
    }

    // Construct the struct definition with generic parameters if any
    if public_fields.is_empty() {
        format!("pub struct {}{} {{}}", name, generic_params)
    } else {
        format!(
            "pub struct {}{} {{\n{}\n}}",
            name,
            generic_params,
            public_fields.join("\n")
        )
    }
}

// Process a public enum and return its external interface
fn process_enum(node: &Node, source: &str) -> String {
    // Extract enum name
    let name_node = node.child_by_field_name("name").unwrap();
    let name = name_node.utf8_text(source.as_bytes()).unwrap();

    // We only process public enums now
    let prefix = "pub ";

    // Get the variant list if it exists (it's called "body" in the AST)
    let variant_list_node = node.child_by_field_name("body");

    // If there's no variant list, return an empty enum
    if variant_list_node.is_none() {
        return format!("{}enum {} {{}}", prefix, name);
    }

    let variant_list_node = variant_list_node.unwrap();

    // Collect variants
    let mut variants: Vec<String> = Vec::new();
    let mut cursor = variant_list_node.walk();

    // Track if we need to add a comma (for all variants except the last one)
    let mut variant_nodes: Vec<Node> = Vec::new();
    for child in variant_list_node.children(&mut cursor) {
        if child.kind() == "enum_variant" {
            variant_nodes.push(child);
        }
    }

    for (_index, child) in variant_nodes.iter().enumerate() {
        // Get the full variant text including type parameters or struct-like fields
        let variant_text = child.utf8_text(source.as_bytes()).unwrap().trim();

        // Always add a comma to match snapshot format
        let with_comma = format!("{},", variant_text);

        variants.push(format!("    {}", with_comma));
    }

    // Construct the enum definition
    if variants.is_empty() {
        format!("{}enum {} {{}}", prefix, name)
    } else {
        format!("{}enum {} {{\n{}\n}}", prefix, name, variants.join("\n"))
    }
}

// Process a public constant and return its definition
fn process_const(node: &Node, source: &str) -> String {
    // Extract the entire constant declaration
    let const_text = node.utf8_text(source.as_bytes()).unwrap();

    // Return the constant declaration as is
    const_text.to_string()
}

// Process an impl block and extract public methods
fn process_impl(node: &Node, source: &str) -> Option<(String, String)> {
    // Extract the type name this impl is for
    let type_node = node.child_by_field_name("type")?;
    let type_name = type_node.utf8_text(source.as_bytes()).unwrap();

    // Get the implementation body
    let body_node = node.child_by_field_name("body")?;

    // Collect public methods
    let mut public_methods = Vec::new();
    let mut cursor = body_node.walk();

    for child in body_node.children(&mut cursor) {
        if child.kind() == "function_item" && is_public(&child, source) {
            // Extract the method text
            let method_text = child.utf8_text(source.as_bytes()).unwrap();

            // Check if it contains "async fn"
            let is_async = method_text.contains("async fn");

            // Get the method signature
            let name_node = child.child_by_field_name("name")?;
            let name = name_node.utf8_text(source.as_bytes()).unwrap();

            // Check for generic type parameters
            let mut generic_params = String::new();
            // Look for the type_parameters node which contains generic parameters
            for type_params_node in child.children(&mut child.walk()) {
                if type_params_node.kind() == "type_parameters" {
                    generic_params = type_params_node
                        .utf8_text(source.as_bytes())
                        .unwrap()
                        .to_string();
                    break;
                }
            }

            // Get the parameters
            let mut params = Vec::new();
            let parameters_node = child.child_by_field_name("parameters")?;
            let mut param_cursor = parameters_node.walk();

            // First check if this method has a self parameter
            let has_self_param = parameters_node
                .children(&mut parameters_node.walk())
                .any(|param| param.kind() == "self_parameter");

            // If it has a self parameter, add it first
            if has_self_param {
                // Try to find the specific self parameter to get accurate text
                let self_text = parameters_node
                    .children(&mut parameters_node.walk())
                    .find(|param| param.kind() == "self_parameter")
                    .map(|param| param.utf8_text(source.as_bytes()).unwrap().to_string())
                    .unwrap_or("&self".to_string()); // Default to &self if not found

                params.push(self_text);
            }

            // Add the rest of the parameters
            for param in parameters_node.children(&mut param_cursor) {
                if param.kind() == "parameter" {
                    let param_text = param.utf8_text(source.as_bytes()).unwrap();
                    params.push(param_text.to_string());
                }
            }

            // Get the return type if any
            let mut return_type = String::new();
            if let Some(return_node) = child.child_by_field_name("return_type") {
                return_type = format!(" -> {}", return_node.utf8_text(source.as_bytes()).unwrap());
            }

            // Construct the method signature
            let method_sig = if is_async {
                format!(
                    "    pub async fn {}{}{};",
                    name,
                    generic_params,
                    format!("({}){}", params.join(", "), return_type)
                )
            } else {
                format!(
                    "    pub fn {}{}{};",
                    name,
                    generic_params,
                    format!("({}){}", params.join(", "), return_type)
                )
            };

            public_methods.push(method_sig);
        }
    }

    // If no public methods, return None
    if public_methods.is_empty() {
        return None;
    }

    // Create the impl block
    let impl_block = format!("impl {} {{\n{}\n}}", type_name, public_methods.join("\n"));

    Some((type_name.to_string(), impl_block))
}

// Process a public function and return its signature
fn process_function(node: &Node, source: &str) -> String {
    // Extract the function declaration text
    let func_text = node.utf8_text(source.as_bytes()).unwrap();

    // Check if it contains "async fn"
    let is_async = func_text.contains("async fn");

    // Get the function name
    let name_node = node.child_by_field_name("name").unwrap();
    let name = name_node.utf8_text(source.as_bytes()).unwrap();

    // Check for generic type parameters
    let mut generic_params = String::new();
    // Look for the type_parameters node which contains generic parameters
    for type_params_node in node.children(&mut node.walk()) {
        if type_params_node.kind() == "type_parameters" {
            generic_params = type_params_node
                .utf8_text(source.as_bytes())
                .unwrap()
                .to_string();
            break;
        }
    }

    // Get the parameters
    let mut params = Vec::new();
    let parameters_node = node.child_by_field_name("parameters").unwrap();
    let mut param_cursor = parameters_node.walk();

    for param in parameters_node.children(&mut param_cursor) {
        if param.kind() == "parameter" {
            let param_text = param.utf8_text(source.as_bytes()).unwrap();
            params.push(param_text.to_string());
        }
    }

    // Get the return type if any
    let mut return_type = String::new();
    if let Some(return_node) = node.child_by_field_name("return_type") {
        return_type = format!(" -> {}", return_node.utf8_text(source.as_bytes()).unwrap());
    }

    // Construct the function signature
    if is_async {
        format!(
            "pub async fn {}{}{};",
            name,
            generic_params,
            format!("({}){}", params.join(", "), return_type)
        )
    } else {
        format!(
            "pub fn {}{}{};",
            name,
            generic_params,
            format!("({}){}", params.join(", "), return_type)
        )
    }
}

// Process a public module declaration
fn process_module(node: &Node, source: &str) -> String {
    // Extract the module name
    let name_node = node.child_by_field_name("name").unwrap();
    let name = name_node.utf8_text(source.as_bytes()).unwrap();

    // Return just the module declaration
    format!("pub mod {};", name)
}

// Process a public type alias
fn process_type_alias(node: &Node, source: &str) -> String {
    // Extract the entire type alias declaration
    let type_text = node.utf8_text(source.as_bytes()).unwrap();

    // Return the type alias declaration as is
    type_text.to_string()
}

// Process a public trait definition
fn process_trait(node: &Node, source: &str) -> String {
    // Extract the trait name
    let name_node = node.child_by_field_name("name").unwrap();
    let name = name_node.utf8_text(source.as_bytes()).unwrap();

    // Get the trait body
    let body_node = node.child_by_field_name("body").unwrap();

    // Collect trait methods
    let mut methods = Vec::new();
    let mut cursor = body_node.walk();

    for child in body_node.children(&mut cursor) {
        // In trait definitions, method signatures appear as function_signature_item
        if child.kind() == "function_signature_item" {
            // Extract the entire method signature
            let signature_text = child.utf8_text(source.as_bytes()).unwrap().trim();
            methods.push(format!("    {}", signature_text));
        }
    }

    // Construct the trait definition
    if methods.is_empty() {
        format!("pub trait {} {{}}", name)
    } else {
        format!("pub trait {} {{\n{}\n}}", name, methods.join("\n"))
    }
}

// Process a public use declaration
fn process_use_declaration(node: &Node, source: &str) -> String {
    // Extract the use declaration text
    let use_text = node.utf8_text(source.as_bytes()).unwrap();

    // Return the use declaration as is
    use_text.to_string()
}
