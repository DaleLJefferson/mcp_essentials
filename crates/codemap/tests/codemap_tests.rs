use codemap::codemap;
use insta::assert_snapshot;

#[test]
fn test_everything() {
    let source_code = include_str!("../test_assets/everything.rs");

    let result = codemap(source_code);

    assert_snapshot!(result);
}
