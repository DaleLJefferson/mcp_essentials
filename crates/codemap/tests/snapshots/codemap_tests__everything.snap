---
source: crates/codemap/tests/codemap_tests.rs
expression: result
---
pub struct Simple {
    pub public_field: i32
}

pub struct PublicPrivatePublic {
    pub public_field: i32,
    pub pub_field: String
}

pub enum PublicEnum {
    Variant1,
    Variant2
}

pub const CONSTANT: i32 = 42;

pub struct ImplStruct;

impl ImplStruct {
    pub fn public_method(param: i32) -> i32;
}

pub fn public_function(param: i32) -> i32;
